core/src/operations/decorators/mod.rs

```diff
@@ -0,0 +1,48 @@
+mod advice;
+use crate::utils::collections::Vec;
+pub use advice::AdviceInjector;
+use core::fmt;
+
+// DECORATORS
+// ================================================================================================
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum Decorator {
+    /// Injects zero or more values at the head of the advice tape as specified by the injector.
+    /// This operation affects only the advice tape, but has no effect on other VM components
+    /// (e.g., stack, memory), and does not advance VM clock.
+    Advice(AdviceInjector),
+}
+
+impl fmt::Display for Decorator {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Advice(injector) => write!(f, "advice({})", injector),
+        }
+    }
+}
+
+pub type DecoratorMap = Vec<(usize, Decorator)>;
+
+pub struct DecoratorIterator<'a> {
+    decorator_map: &'a DecoratorMap,
+    idx: usize,
+}
+
+impl<'a> DecoratorIterator<'a> {
+    pub fn new(decorator_map: &'a DecoratorMap) -> Self {
+        Self {
+            decorator_map,
+            idx: 0,
+        }
+    }
+
+    pub fn next(&mut self, pos: usize) -> Option<&Decorator> {
```

Comment:
You could try to add `#[inline(always)]` here to see if this speeds things up.

Reply:
It does speed things. Brought it back to parity with the earlier approach.

Reply:
Great!